packages/@aws-cdk/aws-cognito-identitypool/lib/identitypool-user-pool-authentication-provider.ts

```diff
@@ -0,0 +1,120 @@
+import { randomBytes } from 'crypto';
+import {
+  CfnUserPool,
+  IUserPool,
+  UserPoolClient,
+  UserPoolClientProps,
+} from '@aws-cdk/aws-cognito';
+import {
+  Construct, Node,
+} from 'constructs';
+import { IIdentityPool } from './identitypool';
+
+/**
+ * Represents the concept of a User Pool Authentication Provider.
+ * You use user pool authentication providers to configure User Pools
+ * and User Pool Clients for use with Identity Pools
+ */
+export interface IUserPoolAuthenticationProvider {
+  /**
+   * The method called when a given User Pool Authentication Provider is added
+   * (for the first time) to an Identity Pool.
+   */
+  bind(
+    scope: Construct,
+    identityPool: IIdentityPool,
+    options?: UserPoolAuthenticationProviderBindOptions
+  ): UserPoolAuthenticationProviderBindConfig;
+}
+
+/**
+ * Props for the User Pool Authentication Provider
+ */
+export interface UserPoolAuthenticationProviderProps extends UserPoolClientProps, UserPoolAuthenticationProviderBindOptions {}
```

Comment:
```suggestion
export interface UserPoolAuthenticationProviderProps {
  readonly disableServerSideTokenCheck?: boolean;
  readonly userPool: IUserPool;
  readonly userPoolClient: IUserPoolClient;
}
```

We need to provide a way for the user to provide their own user pool client. And rather than exposing all the client props and creating one for them let's just ask them to provide a client. It's easy enough for them to just go `userPool.addClient()` if they don't have one already.

Reply:
What do you think about making `userPoolClient` optional and if it's not there just doing `userPool.addClient()` in the constructor?

Reply:
A similar strategy is used [here](https://github.com/aws/aws-cdk/blob/47efcf89828a72be5f1ef7d60efba578183c3fde/packages/%40aws-cdk/aws-apigatewayv2-authorizers/lib/http/user-pool.ts#L54)

Reply:
That's fine too.

Reply:
Ok, will push shortly

Reply:
@corymhall Pushed changes, just waiting for tests to finish then I think we're good to go